src/Build/Logging/BinaryLogger/BinaryLogReplayEventSource.cs

```diff
@@ -0,0 +1,63 @@
+﻿using System;
+using System.IO;
+using System.IO.Compression;
+using Microsoft.Build.Framework;
+
+namespace Microsoft.Build.Logging
+{
+    /// <summary>
+    /// Provides a method to read a binary log file (*.binlog) and replay all stored BuildEventArgs
+    /// by implementing IEventSource and raising corresponding events.
+    /// </summary>
+    public class BinaryLogReplayEventSource : EventArgsDispatcher
+    {
+        /// <summary>
+        /// Read the provided binary log file and raise corresponding events for each BuildEventArgs
+        /// </summary>
+        /// <param name="sourceFilePath">The full file path of the binary log file</param>
+        public void Replay(string sourceFilePath)
+        {
+            using (var stream = new FileStream(sourceFilePath, FileMode.Open))
+            {
+                var gzipStream = new GZipStream(stream, CompressionMode.Decompress, leaveOpen: true);
+                var binaryReader = new BinaryReader(gzipStream);
+
+                byte fileFormatVersion = binaryReader.ReadByte();
```

Comment:
Eventually, it would be nice to error here if the file formats do not match. Or alternatively, keep all the readers/writers for each file version and delegate to the correct one. This might be a common case. I suspect people would want to switch their CI builds to log the binary format rather than the text one, and then you'd have cases where people are assigned to investigate some old failed build. But this can be done whenever fileFormatVersion=2 comes along.